lab_2_tokenize_by_bpe/main.py

```diff
@@ -26,53 +40,147 @@ def collect_frequencies(
     :return: dictionary in the form of <token pair: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict):
+        return None
+
+    pair_dict = {}
+
+    for word in word_frequencies:
+        for index in range(len(word) - 1):
+            pair = (word[index], word[index + 1])
+            if pair not in pair_dict:
+                pair_dict[pair] = 0
+            pair_dict[pair] += word_frequencies[word]
+
+    return pair_dict
 
 
 def merge_tokens(
-    word_frequencies: dict[tuple[str, ...], int], pair: tuple[str, str]
+        word_frequencies: dict[tuple[str, ...], int], pair: tuple[str, str]
 ) -> dict[tuple[str, ...], int] | None:
     """
     Updates word frequency dictionary by replacing a pair of token with a merged one
     :param word_frequencies: dictionary in the form of <preprocessed word: number of occurrences>
     :param pair: a pair of tokens to be merged
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict) or not isinstance(pair, tuple):
+        return None
+
+    word_freq_updated = word_frequencies.copy()
+
+    for word in word_frequencies:
+        new_word = []
+        if pair[0] in word and pair[1] in word:
+            for index, token in enumerate(word):
+                if token == pair[1] and word[index - 1] == pair[0]:
+                    pass
+                elif token != pair[0]:
+                    new_word.append(token)
+                elif token == pair[0] and word[index + 1] != pair[1]:
+                    new_word.append(token)
+                elif token == pair[0] and word[index + 1] == pair[1]:
+                    new_word.append(pair[0] + pair[1])
+
+            value = word_freq_updated.pop(word)
+            word_freq_updated[tuple(new_word)] = value
+
+    return word_freq_updated
 
 
 def train(
-    word_frequencies: dict[tuple[str, ...], int] | None, num_merges: int
+        word_frequencies: dict[tuple[str, ...], int] | None, num_merges: int
 ) -> dict[tuple[str, ...], int] | None:
     """
     Creates required number of new tokens by merging existing ones
     :param word_frequencies: dictionary of a kind <preprocessed word: number of occurrences>
     :param num_merges: required number of new tokens
     :return: dictionary in the form of <preprocessed word: number of occurrences>
     """
+    if not isinstance(word_frequencies, dict) or not isinstance(num_merges, int):
+        return None
+
+    pair_dict = count_tokens_pairs(word_frequencies)
+    if pair_dict is None:
+        return None
+
+    for _ in range(num_merges):
+        if pair_dict != {}:
+            max_value = max(pair_dict.values())
+            pair_list = [key for key, value in pair_dict.items() if value == max_value]
+
+            pair_list.sort()
+            pair_to_merge = max(pair_list, key=lambda x: len(x[0] + x[1]))
```

Comment:
since x is the element from pair_list( a pair (tuple with two strings)) then x[0] is the first element (str) in that pair and x[1] is the second element (str) in that pair => x[0] + x[1] is one string formed after concatenation of both elements of a pair => it's a merged pair